homeassistant/components/upcloud/__init__.py

```diff
@@ -52,65 +64,199 @@
+            )
+        }
+
+
+@dataclasses.dataclass
+class UpCloudHassData:
+    """Home Assistant UpCloud runtime data."""
+
+    coordinators: Dict[str, UpCloudDataUpdateCoordinator] = dataclasses.field(
+        default_factory=dict
+    )
+    scan_interval_migrations: Dict[str, int] = dataclasses.field(default_factory=dict)
+
+
+async def async_setup(hass: HomeAssistantType, config) -> bool:
+    """Set up UpCloud component."""
+    domain_config = config.get(DOMAIN)
+    if domain_config:
+        _LOGGER.warning(
+            "Loading upcloud via top level config is deprecated and no longer "
+            "necessary as of 0.117. Please remove it from your YAML configuration."
+        )
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data={
+                    CONF_USERNAME: domain_config[CONF_USERNAME],
+                    CONF_PASSWORD: domain_config[CONF_PASSWORD],
+                },
+            )
+        )
+
+        if domain_config[CONF_SCAN_INTERVAL]:
+            hass.data[DATA_UPCLOUD] = UpCloudHassData()
+            hass.data[DATA_UPCLOUD].scan_interval_migrations[
+                domain_config[CONF_USERNAME]
+            ] = domain_config[CONF_SCAN_INTERVAL]
+
+    return True
+
+
+def _config_entry_update_signal_name(config_entry: ConfigEntry) -> str:
+    """Get signal name for updates to a config entry."""
+    return CONFIG_ENTRY_UPDATE_SIGNAL_TEMPLATE.format(config_entry.unique_id)
+
+
+async def _async_signal_options_update(
+    hass: HomeAssistantType, config_entry: ConfigEntry
+) -> None:
+    """Signal config entry options update."""
+    async_dispatcher_send(
+        hass, _config_entry_update_signal_name(config_entry), config_entry
+    )
 
-    manager = upcloud_api.CloudManager(username, password)
+
+async def async_setup_entry(hass: HomeAssistantType, config_entry: ConfigEntry) -> bool:
+    """Set up the UpCloud config entry."""
+
+    manager = upcloud_api.CloudManager(
+        config_entry.data[CONF_USERNAME], config_entry.data[CONF_PASSWORD]
+    )
 
     try:
-        manager.authenticate()
-        hass.data[DATA_UPCLOUD] = UpCloud(manager)
+        await hass.async_add_executor_job(manager.authenticate)
     except upcloud_api.UpCloudAPIError:
-        _LOGGER.error("Authentication failed")
+        _LOGGER.error("Authentication failed", exc_info=True)
         return False
-
-    def upcloud_update(event_time):
-        """Call UpCloud to update information."""
-        _LOGGER.debug("Updating UpCloud component")
-        hass.data[DATA_UPCLOUD].update()
-        dispatcher_send(hass, SIGNAL_UPDATE_UPCLOUD)
+    except requests.exceptions.RequestException as err:
+        _LOGGER.error("Failed to connect", exc_info=True)
+        raise ConfigEntryNotReady from err
+
+    upcloud_data = hass.data.setdefault(DATA_UPCLOUD, UpCloudHassData())
```

Comment:
Shouldn't this be separated by `config_entry.entry_id`? Now you can only support one instance.

Reply:
`UpCloudHassData` contains a dict keyed by usernames with data update coordinator values, and usernames are config entry unique ids, so we do support multiple accounts/entries.